src/Compiler/Microsoft.CodeAnalysis.Razor/src/RenderModeTagHelperDescriptorProvider.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+using System;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Components;
+using static Microsoft.AspNetCore.Razor.Language.CommonMetadata;
+
+namespace Microsoft.CodeAnalysis.Razor;
+
+internal class RenderModeTagHelperDescriptorProvider : ITagHelperDescriptorProvider
+{
+    private static TagHelperDescriptor? s_refTagHelper;
+
+    // Run after the component tag helper provider
+    public int Order { get; set; } = 1000;
+
+    public RazorEngine? Engine { get; set; }
+
+    public void Execute(TagHelperDescriptorProviderContext context)
+    {
+        if (context == null)
+        {
+            throw new ArgumentNullException(nameof(context));
+        }
+
+        var compilation = context.GetCompilation();
+        if (compilation == null)
+        {
+            return;
+        }
+
+        var iComponentRenderMode = compilation.GetTypeByMetadataName(ComponentsApi.IComponentRenderMode.FullTypeName);
```

Comment:
This lookup means that if the type is defined twice then the feature fails. Previsouly @jjonescz did work to do more specific look up types that should come from the ASP.NET runtime. Can we re-use that here?

Reply:
I assume we can just use `GetTypesByMetadata` instead and pick a winner. @jjonescz Link to that code?

Reply:
Yes, that's basically what I did - check `GetTypesByMetadataName` - but only if it comes from `Microsoft.AspNetCore.Components.dll` and is `public`.

https://github.com/dotnet/razor/blob/7530beab777ef0eb50fcd193a142e57276c39474/src/Compiler/Microsoft.NET.Sdk.Razor.SourceGenerators/RazorSourceGenerator.cs#L223-L232

Note that creating compilation for references is there only because it's an incremental source generator, shouldn't be needed here.

Reply:
Ah ok. So this isn't going to work right now, because we don't have the actual references to pull from, so we can't do this just yet. 

@jaredpar `GetTypeByMetadata` only fails if there are multiple external references that contain the requested type right, preferencing source if it's there, so I don't see this being a huge issue? https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.compilation.gettypebymetadataname?view=roslyn-dotnet-4.6.0

Reply:
It's a bit of a corner case.